app/src/game-engine/local-engine/rules/CastlingRule.ts

```diff
@@ -0,0 +1,171 @@
+import { Rule, RuleEvaluationResult } from "../Rule";
+import { GameState } from "../GameState";
+import { Move } from "../extensions/Move";
+import { ISquare } from "../../ISquare";
+import { PieceType, PieceColor } from "../../IPiece";
+
+enum castlingFlags {
+    castled,
+    kingMoved,
+    shortCastlingRookMoved,
+    longCastlingRookMoved
+}
+
+export class CastlingRule extends Rule {
+
+    private castlingPieces:string[] = ["King","Rook"];
+    // next array strictly depends on the indexes for PieceColor, 0 for Black and 1 for White.
+    // may need to remove this dependency ?
+    private piecesFixedPositions = [
+        {
+            initialRookShortCastling: { row: 0, column: 7 },
+            initialRookLongCastling: { row: 0, column: 0 },
+            finalKingShortCastling: { row: 0, column: 6 },
+            finalKingLongCastling: { row: 0, column: 2 },
+            finalRookShortCastling: { row: 0, column: 5 },
+            finalRookLongCastling: { row: 0, column: 3 }
+        },
+        {
+            initialRookShortCastling: { row: 7, column: 7 },
+            initialRookLongCastling: { row: 7, column: 0 },
+            finalKingShortCastling: { row: 7, column: 6 },
+            finalKingLongCastling: { row: 7, column: 2 },
+            finalRookShortCastling: { row: 7, column: 5 },
+            finalRookLongCastling: { row: 7, column: 3 }
+        }
+    ]
+    private allowedKingCols = [6,2]; // cols that shoot the castling king movement
+
+    public evaluate(move: Move, state: GameState): RuleEvaluationResult {
+        let movingPiece = state.board.getPiece(move.source);
+        if (!!movingPiece){
+            if (!this.castlingPieces.includes(movingPiece.type)) {
+                return this.nextOrInvalidResult(move, state);
+            }
+            let pieceColor = movingPiece.color;
+            // if next condition is done, player can't castle anymore
+            if (state.castlingFlags[pieceColor][castlingFlags.castled] || state.castlingFlags[pieceColor][castlingFlags.kingMoved]){
+                return this.nextOrInvalidResult(move, state);
+            }
+            return movingPiece.type === PieceType.King ? this.proceedForKing(move,state,pieceColor): this.proceedForRoot(move, state, pieceColor);
+        }
+        return { valid: false };
+    }
```

Comment:
I would simplify this by accepting castling only from a King move (as stated in official chess rules).

Reply:
The proceed for root is only there to update the flags.. it doesn't mean to make the castling movement. If you move a rook once you can't casttle that side anymore